elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py

```diff
@@ -0,0 +1,30 @@
+from typing import Any
+
+import torch
+
+from ._quantize_to_fixed_point import quantize
+from ._two_complement_fixed_point_config import FixedPointConfigV2
+
+
+class RoundToFixedPoint(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx: Any, *args: Any, **kwargs: Any) -> torch.Tensor:
+        if len(args) != 3:
+            raise TypeError(
+                "apply() takes exactly two arguments "
+                "(x: torch.Tensor, config: FixedPointConfig)"
+            )
+        x: torch.Tensor = args[0]
+        fxp_config: FixedPointConfigV2 | None = args[1]
+        ctx.grad_fxp_config: FixedPointConfigV2 | None = args[2]
+        if fxp_config is None:
```

Comment:
This instead might be an actual problem. The RoundToFixedPoint function not rounding to fixed point but instead acting as an identity seems like a Liskov violation to me. I'd never expect this function to not somehow round a value. We could actually raise an error in case no fxp config is given.

Reply:
While I agree with your take that this function should from the naming scheme never do this, I would really like to have a chance to inherit no FxpConfig to be able to not round something. But I have no idea how to do this easily. Do you have an idea?

Reply:
Not sure, what you mean with "inherit" in this context. But here's how Julian or me would probably do it.
Instead of injecting a boolean parameter, we inject the function itself. Somewhere you will call `RoundToFixedPoint.apply(x, None)` in a layer. Let's say that layer is called `Conv`. So `Conv` looks like this

```python
class Conv(torch.nn.Conv):
  def __init__(self, fxp_config: FxpConfig | None):
    super().__init__()
    self._fxp_config = fxp_config

  def forward(self, x):
    weights = RoundToFixedPoint.apply(self.weights, self._fxp_config)
    return torch.nn.functional.conv(x, weights)
```

Instead you write

```python
class Conv(torch.nn.Conv):
  def __init__(self, quantize: Callable[[Tensor], Tensor]):
    super().__init__()
    self._quantize = quantize

  def forward(self, x):
    weights = quantize(self.weights, self._fxp_config)
    return torch.nn.functional.conv(x, weights)

def round_to_fixed_point(x: Tensor, fxp_config: FxpConfig) -> Tensor:
  return RoundToFixedPoint.apply(x, fxp_config)

from functools import partial

my_config = FxpConfig(total_bits=4, frac_bits=3)
c = Conv(quantize=partial(
        round_to_fixed_point, 
        fxp_config=my_conf))
c_without_quant = Conv(quantize=lambda x: x)
```